compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2466,11 +2466,25 @@ impl<'a> Parser<'a> {
         if !self.eat_keyword(kw::In) {
             self.error_missing_in_for_loop();
         }
+        let expr_token = self.token.clone();
         self.check_for_for_in_in_typo(self.prev_token.span);
         let expr = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
 
         let pat = self.recover_parens_around_for_head(pat, begin_paren);
 
+        // Recover from missing expression in `for` loop
+        if matches!(expr_token.kind, token::OpenDelim(token::Delimiter::Brace))
```

Comment:
We don't need to use `expr_token` here, we should be able to check that the `expr`'s kind is a `ExprKind::Block` or something like that.